plugins/hls-cabal-fmt-plugin/src/Ide/Plugin/CabalFmt.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Ide.Plugin.CabalFmt where
+
+import           Control.Lens
+import           Control.Monad.IO.Class
+import qualified Data.Text               as T
+import           Development.IDE         hiding (pluginHandlers)
+import           Ide.PluginUtils
+import           Ide.Types
+import           Language.LSP.Types      as J
+import qualified Language.LSP.Types.Lens as J
+import           Prelude                 hiding (log)
+import           System.Directory
+import           System.Exit
+import           System.FilePath
+import           System.Process
+
+data Log
+  = LogProcessInvocationFailure Int
+  | LogReadCreateProcessInfo String [String]
+  | LogInvalidInvocationInfo
+  deriving (Show)
+
+instance Pretty Log where
+  pretty = \case
+    LogProcessInvocationFailure code -> "Invocation of cabal-fmt failed with code" <+> pretty code
+    LogReadCreateProcessInfo stdErrorOut args -> vcat
+      $ ["Invocation of cabal-fmt with arguments" <+> pretty args]
+      ++ ["failed with standard error:" <+> pretty stdErrorOut | not (null stdErrorOut)]
+    LogInvalidInvocationInfo -> "Invocation of cabal-fmt with range was called but is not supported."
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId =
+  (defaultCabalPluginDescriptor plId)
+    { pluginHandlers = mkFormattingHandlers (provider recorder)
+    }
+
+-- | Formatter provider of cabal fmt.
+-- Formats the given source in either a given Range or the whole Document.
+-- If the provider fails an error is returned that can be displayed to the user.
+provider :: Recorder (WithPriority Log) -> FormattingHandler IdeState
+provider recorder _ (FormatRange _) _ _ _ = do
+  logWith recorder Info LogInvalidInvocationInfo
+  pure $ Left (ResponseError InvalidRequest "You cannot format a text-range using cabal-fmt." Nothing)
```

Comment:
Hmm, do we not have a better way to have a formatting provider that only supports whole buffer formatting? I think some of our others have this limitation too :thinking:

Reply:
I checked all formatter plugins (stylish-haskell, ormolu, fourmolu, brittany, floskell), everyone seems to handle ranges :/

Reply:
I think maybe we could do something clever with dynamic registration to say we only support it in certain circumstances... but that seems hard. This seems okay I guess. Worth a comment, though.

Reply:
What kind of comment are you thinking here?